### PR TITLE
[Manager] Manager use TensorPool for Weights

### PR DESCRIPTION
**Pull request by kparichay**
- This patch adds bugfixes for layers:
1. acti_func in-place support bug
2. batch normalization layer weight name bug
3. loss layers not using tensor reference bug
- This patch updates manager to use TensorPool for its weights and
corresponding weights. Corresponding changes to Var_Grad/Weights and
Tensors are also added.
- setBatchSize() now takes dimensions from user, and then updates the
provides these updates dimensions to the manager. We can possibly remove
setBatch(RunLayerContext), which will be finalized in the next commit.
- Use TensorPool for gradients of the weights.
- This patch makes tensor list in tensorpool ordered.
This ensures that the requested tensors are initialized in the order of
the their requests which is by the order of the sorted graph.
- This patch adds fix when rebase to the main branch.
Due to significant changes to the main branch, certain patches have
been taken from future PRs:
  - from #1539, commit b58c72a has been
cherry picked to solve the issue of tensor shape related with batch size
changes
  - from #1539, memory pool to start giving tokens from 1 instead of 0. token 0 is
treated at a non-requested tensor memory

  These patches might seem a bit out of the way but instead of taking
partial functionality, either full commits have been cherry-picked here
or minor functionality has been manually imported.

See Also #1127

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: nntrainer/graph/network_graph.cpp [hunk 1]
```diff
@@ -422,11 +422,34 @@ int NetworkGraph::realizeGraph() {
...
+    tensor_manager->deallocateTensors();
+
   for (auto iter = cbegin(); iter != cend(); iter++) {
     (*iter)->setBatch(batch_size);
+    if ((*iter)->isRunContextAvailable()) {
+      const RunLayerContext &context = (*iter)->getRunContext();
+      // resize tensors spec
+      for (unsigned int idx = 0; idx < context.getNumTensors(); idx++) {
+        auto const &ts = context.getTensor(idx);
+        tensor_manager->setBatchSize(ts.getName(), ts.getDim().batch());
```
Comment (zhoonit) [inline nntrainer/graph/network_graph.cpp:441]: Q. what's the difference between this and line 449?

Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:441]: L441 sets the batch size for the individual tensors by name - this covers the tensors requested by the layers.
L449 sets batch size for the tensors whose behaviors are known - weights, gradients, inputs, outputs, and derivatives.

Comment (zhoonit) [reply nntrainer/graph/network_graph.cpp:441]: okay

File: nntrainer/tensor/manager.cpp [hunk 2]
```diff
@@ -380,14 +364,7 @@ void Manager::deallocateGradients() {
...
 void Manager::initializeGradients() {
-  if (LAYER_V2) {
-    for (auto &w : weights_v2) {
-      w->initializeGradient();
-      // auto exec_order = tensor_exec_order[w.getName()];
-      // tensor_map(&w->getGradientRef(), requestMemory(w.getDim().size(),
-      //       std::get<1>(exec_order), std::get<2>(exec_order) + 1));
-    }
-  } else {
+  if (!LAYER_V2) {
```
[the remark this answers is not in the document]
Comment (zhoonit) [reply nntrainer/tensor/manager.cpp:367]: Can we delete if(LAYER_V2) ? it make it hard to read code.

**Comment by zhoonit** [comment]
> We can possibly remove setBatch(RunLayerContext)

1. Can we remove setBatch(RunLayerContext) ? this is used for the supporting setting batch after the initialization.
2. Can we remove setBatch(InitLayerContext) ? My answer is yes but seeing this PR, maybe there might be another usage.

File: nntrainer/tensor/tensor_pool.h [hunk 3]
```diff
@@ -151,8 +172,13 @@ class TensorPool {
...
   };
 
-  std::unordered_map<std::string, requestSpec>
-    pool;              /**< list of requested tensors */
+  /**
+   * note: unordered_map is not directly used for pool to ensure initialization
+   * of weights
+   */
+  std::vector<requestSpec> pool; /**< list of requested tensors */
+  std::unordered_map<std::string, unsigned int>
```
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/tensor/tensor_pool.h:180]: `ordered_map` will order by the key itself, but the ordering required here is by the order of insertion.

File: nntrainer/tensor/tensor_pool.cpp [hunk 4]
```diff
@@ -142,9 +147,12 @@ void TensorPool::allocate() {
   mem_pool.allocate();
 
   /** set the pointers using the token for all the tensors */
-  for (auto &entry : pool) {
-    auto &spec = entry.second;
-    spec.tensor->setData(mem_pool.getMemory(spec.token));
+  for (auto &spec : pool) {
+    /** get data for the tensors which were requested */
+    if (spec.token > 0) {
```
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/tensor/tensor_pool.cpp:152]: spec.token == 0 -> that requested tensor does not has memory allocated.

File: nntrainer/tensor/manager.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/tensor/manager.cpp:367]: LAYER_V2 is deleted with #1531. If I remove it here, the rebase will be issue again.

File: nntrainer/tensor/tensor_pool.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (zhoonit) [reply nntrainer/tensor/tensor_pool.h:180]: I see, thanks :)

File: nntrainer/tensor/manager.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (zhoonit) [reply nntrainer/tensor/manager.cpp:367]: Cool. I'll wait.

File: nntrainer/layers/layer_node.h [hunk 5]
```diff
@@ -560,6 +560,18 @@ class LayerNode final : public ml::train::Layer, public GraphNode {
     return *run_context;
   }
 
+  /**
+   * @brief   check if run layer context is available
+   *
+   * @retval  bool true if context is available else false
+   */
+  bool isRunContextAvailable() const {
```
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/layers/layer_node.h:568]: This is needed to determine that if the layer has context already defined, then update the batchsize inside the context. Updated to `isFinalized()` with the rebase.

File: nntrainer/tensor/var_grad.h [hunk 6]
```diff
@@ -346,7 +356,13 @@ class Var_Grad {
    * @note this is can return is the var_grad needs gradient but it not
    * empty
    */
-  bool hasGradient() const { return !grad->empty(); }
+  bool hasGradient() const {
+    if (!grad)
+      return false;
+    if (var->isAllocated())
+      return grad->isAllocated();
```
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/tensor/var_grad.h:363]: `hasGradient()` combines the below questions into one:
- was this var_grad created with gradient
- does the gradient has defined shape
- has the gradient been allocated

There are various scenarios where its used with different concepts, and it would require all these checks independently to be done here. This function wraps all of them, and provides guarantees that if the return value is true, gradient is available and can be used safely.

File: nntrainer/layers/layer_context.cpp [hunk 7]
```diff
@@ -193,6 +203,19 @@ Tensor &RunLayerContext::getTensorGrad(unsigned int idx) {
...
 }
 
+/**
+ * @brief Get the Tensor Grad object
+ *
+ * @param idx Identifier of the tensor
+ * @return Tensor& Reference to the tensor grad tensor
+ */
+const Tensor &RunLayerContext::getTensorGrad(unsigned int idx) const {
+  if (!tensors[idx]->hasGradient())
```
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/layers/layer_context.cpp:213]: There are two checks:
1. was this tensor even requested - that will be done with #1458 using `.at()`
2. was a gradient requested for this tensor (was the request for trainable tensor or regular)

The check here performs the second check. So, this behavior cannot be employed with getTensor() as it only requires the first check.

File: nntrainer/tensor/manager.cpp [hunk 8]
```diff
@@ -809,34 +811,61 @@ Manager::requestWeights(const GraphNode &node,
...
+      );
+    max_exec_order =
+      std::max(max_exec_order,
+               *std::max_element(var_exec_order.begin(), var_exec_order.end()));
+
+    Tensor *grad = nullptr;
+    // TODO: change to enum_class_and
+    if (std::get<2>(ts) /** need gradient */ &&
+        enum_class_or(tspan, TensorLifespan::FORWARD_FUNC_LIFESPAN) !=
+          TensorLifespan::FORWARD_FUNC_LIFESPAN)
```
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/tensor/manager.cpp:854]: It checks if the requested tspan is only FORWARD_FUNC_LIFESPAN. If it is, then need not request gradient for this tensor.

[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/tensor/manager.cpp:854]: But this check is later simplified with https://github.com/nnstreamer/nntrainer/blob/08298dbaa856b055585047013ae1671ed16d76ab/nntrainer/tensor/manager.cpp#L345-L346

File: nntrainer/tensor/manager.cpp [hunk 9]
```diff
@@ -529,9 +506,9 @@ void Manager::allocateInOuts() {
     for (auto &out : outputs_v2) {
       out->allocateVariable();
     }
-    for (auto &t : tensors_v2) {
-      t->allocateVariable();
-    }
+    // for (auto &t : tensors_v2) {
+    //   t->allocateVariable();
+    // }
```
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/tensor/manager.cpp:511]: deleted.

File: nntrainer/tensor/tensor_pool.cpp [hunk 10]
```diff
@@ -57,21 +58,25 @@ Tensor *TensorPool::requestTensor(const TensorDim dim,
  * compatible.
  */
 Tensor *TensorPool::requestPrerequestedTensor(
-  const TensorDim dim, const std::vector<unsigned int> &exec_order,
-
-  TensorLifespan lifespan, const std::string &name) {
-  if (pool.find(name) == pool.end())
+  const TensorDim &dim, const std::vector<unsigned int> &exec_order,
+  TensorLifespan lifespan, const std::string &name,
+  const Tensor::Initializer &init) {
```
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/tensor/tensor_pool.cpp:63]: > Is there any reason to combine requestTensor and requestPrerequestedTensor?

I am not sure what you mean here, but they aren't combined. Did you mean "any reason to NOT combine"?

**Comment by kparichay** [comment]
> > We can possibly remove setBatch(RunLayerContext)
> 1. Can we remove setBatch(RunLayerContext) ? this is used for the supporting setting batch after the initialization.
> 2. Can we remove setBatch(InitLayerContext) ? My answer is yes but seeing this PR, maybe there might be another usage.

The purpose of either of them is to able to allow the user to update the batchsize of the tensors they have requested. 
1. setBatch(RunLayerContext) - this is the easier choice from implementation perspective
2. setBatch(InitLayerContext) - this is easier for the user IMO as user used InitLayerContext to request the tensors in the first place

But only one is needed. So, the other will be removed soon.

File: nntrainer/tensor/manager.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (zhoonit) [reply nntrainer/tensor/manager.cpp:851]: Cool

File: nntrainer/layers/layer_context.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (zhoonit) [reply nntrainer/layers/layer_context.cpp:213]: Yes, I mean https://github.com/nnstreamer/nntrainer/pull/1522#discussion_r717391889 for the getTensors() those checks aren't done.

Call to this function when !grad->allocated() will throw. But call to getTensor() with tensor->allocated() will not throw.

If they don't have a same semantics, it would cause some bugs later.

[the remark this answers is not in the document]
Comment (zhoonit) [reply nntrainer/layers/layer_context.cpp:213]: But this is minor issue, and slightly off the topic of this PR. I'll approve

File: nntrainer/tensor/tensor_pool.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (zhoonit) [reply nntrainer/tensor/tensor_pool.cpp:63]: Sorry, this is about the recent reply in the previous commit, I though it's the same function.
https://github.com/nnstreamer/nntrainer/pull/1489#discussion_r716374783

My comment is that if we need `dim`, `init`  here as an argument?

File: nntrainer/tensor/var_grad.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (zhoonit) [reply nntrainer/tensor/var_grad.h:363]: > > > We can possibly remove setBatch(RunLayerContext)
> > 
> > 
> > 
> > 1. Can we remove setBatch(RunLayerContext) ? this is used for the supporting setting batch after the initialization.
> > 2. Can we remove setBatch(InitLayerContext) ? My answer is yes but seeing this PR, maybe there might be another usage.
> 
> The purpose of either of them is to able to allow the user to update the batchsize of the tensors they have requested.
> 
> 1. setBatch(RunLayerContext) - this is the easier choice from implementation perspective
> 2. setBatch(InitLayerContext) - this is easier for the user IMO as user used InitLayerContext to request the tensors in the first place
> 
> But only one is needed. So, the other will be removed soon.

Got the gist. Thanks.

**Review by zhoonit — APPROVED** [review]
LGTM

File: nntrainer/tensor/tensor_pool.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (zhoonit) [reply nntrainer/tensor/tensor_pool.cpp:63]: Seeing the next PR, it looks reasonable to check those in here just to prevent human error though.

**Comment by zhoonit** [comment]
> > > We can possibly remove setBatch(RunLayerContext)
> > 
> > 
> > 
> > 1. Can we remove setBatch(RunLayerContext) ? this is used for the supporting setting batch after the initialization.
> > 2. Can we remove setBatch(InitLayerContext) ? My answer is yes but seeing this PR, maybe there might be another usage.
> 
> The purpose of either of them is to able to allow the user to update the batchsize of the tensors they have requested.
> 
> 1. setBatch(RunLayerContext) - this is the easier choice from implementation perspective
> 2. setBatch(InitLayerContext) - this is easier for the user IMO as user used InitLayerContext to request the tensors in the first place
> 
> But only one is needed. So, the other will be removed soon.

Got the gist. Thanks.

File: nntrainer/tensor/manager.cpp [hunk 11]
```diff
@@ -780,25 +761,49 @@ void Manager::deinitializeTensors() {
...
-
-    /** usage for its gradient only in calcGradient */
-    tensor_exec_order[gname].push_back(std::get<1>(exec_order));
-
-    /** set tensor lifespan */
-    expandLifespan(vname, TensorLifespan::MAX_LIFESPAN);
-    expandLifespan(gname, TensorLifespan::BACKWARD_FUNC_LIFESPAN);
+  std::vector<unsigned int> var_exec_order({std::get<0>(exec_order),
+                                            std::get<1>(exec_order),
+                                            std::get<2>(exec_order)});
```
Comment (jijoongmoon) [inline nntrainer/tensor/manager.cpp:761]: I guess, 0 : forwarding, 1: calGradient, 2: calDerivatives.

File: nntrainer/graph/network_graph.cpp [hunk 12]
```diff
@@ -430,12 +430,15 @@ void NetworkGraph::setBatchSize(unsigned int batch_size) {
...
   for (auto iter = cbegin(); iter != cend(); iter++) {
     (*iter)->setBatch(batch_size);
-    const InitLayerContext &init_context = (*iter)->getInitContext();
-    // resize tensors spec
-    for (auto const &ts : init_context.getTensorsSpec()) {
-      tensor_manager->setBatchSize(std::get<3>(ts), batch_size);
-      tensor_manager->setBatchSize(std::get<3>(ts) + Var_Grad::grad_suffix,
-                                   batch_size);
+    if ((*iter)->isRunContextAvailable()) {
+      const RunLayerContext &context = (*iter)->getRunContext();
```
Comment (jijoongmoon) [inline nntrainer/graph/network_graph.cpp:437]: When is the RunContext available? after finalizing the layer? and i wonder when this setBatchSize called? and I also wonder the difference between first setBatch and the other. 
The first one is only for the input and output .. and the the other is for the tensor which is used during calculation?

**Review by jijoongmoon — APPROVED** [review]
LGTM except for minor comments.

File: nntrainer/layers/layer_context.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/layers/layer_context.cpp:213]: Assuring that the tensor is allocated is part of the design itself. For all the tensors given out by the RunLayerContext, the assumption is that graph.allocate() will set and allocate them. So, the checks are not required by design.

File: nntrainer/tensor/tensor_pool.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/tensor/tensor_pool.cpp:63]: init is needed as conflicting initializations will produce wrong results.
dim is added to prevent human error.

File: nntrainer/tensor/manager.cpp (hunk 11 above)
Comment (kparichay) [reply nntrainer/tensor/manager.cpp:761]: Yes. my bad, comments are added after 2 PRs.

File: nntrainer/graph/network_graph.cpp (hunk 12 above)
Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:437]: > When is the RunContext available? after finalizing the layer?

Yes, its right. In this PR in last commit, `isRunContextAvailable` is renamed to `isFinalized()`. 

> and i wonder when this setBatchSize called? and I also wonder the difference between first setBatch and the other.

`setBatchSize()` is called with every call to train/validation/inference. Although the memory is not allocated at that moment, but the size/shapes are set (in case of inference, memory can be allocated as well). In case of first setBatch, memory wont be allocated, and this code will always run. With subsequent calls, if the batch size is same, this function can be a no-op.

> The first one is only for the input and output .. and the the other is for the tensor which is used during calculation?

L441 sets the batch size for the individual tensors by name - this covers the tensors requested by the layers.
L449 sets batch size for the tensors whose behaviors are known - weights, gradients, inputs, outputs, and derivatives.

File: nntrainer/tensor/manager.cpp [hunk 13]
```diff
@@ -814,34 +808,61 @@ Manager::requestWeights(const GraphNode &node,
...
           tspan, TensorLifespan::FORWARD_FUNC_LIFESPAN))
-      tensor_exec_order[vname].push_back(std::get<0>(exec_order));
+      var_exec_order.push_back(std::get<0>(exec_order));
 
     /** usage for tensors gradient in backwarding */
     if (enum_class_logical_and<TensorLifespan>(
           tspan, TensorLifespan::BACKWARD_FUNC_LIFESPAN)) {
-      tensor_exec_order[vname].push_back(std::get<1>(exec_order));
-      tensor_exec_order[gname].push_back(std::get<1>(exec_order));
+      var_exec_order.push_back(std::get<1>(exec_order));
```
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/tensor/manager.cpp:829]: > I wonder if the life span is backward, then does the forwarding is also included?

if the lifespan is backward, that tensor will not be available in forward.

> I mean.. what kind of tensors are included in tensors_spec? Do we assume the none of the tensor which is available at backwording only?

consider batch normalization and convolution layers. The temporary tensors they use during forward and backward operation are requested with FORWARD and BACKWARD lifespans correspondingly.
